simple/stats/stat_var_hierarchy_generator.py

```diff
@@ -12,63 +12,245 @@
+    return replace(self, pvs=pvs)
+
+
+class SVG:
+
+  def __init__(self, svg_id: str, svg_name: str) -> None:
+    self.svg_id = svg_id
+    self.svg_name = svg_name
+
+    # Using dict instead of sets below to maintain insertion order.
+    self.sv_ids: dict[str, bool] = {}
+    self.parent_svg_ids: dict[str, bool] = {}
+    self.child_svg_ids: dict[str, bool] = {}
+
+    self.parent_svgs_processed: bool = False
+    self.has_prop_without_val: bool = False
+    self.sample_sv: SVPropVals | None = None
+
+  def from_sv(sv: SVPropVals, insert_sv_id: bool = False) -> "SVG":
+    svg = SVG(svg_id=sv.gen_svg_id(), svg_name=sv.gen_svg_name())
+    svg.sample_sv = sv
+    if insert_sv_id:
+      svg.sv_ids[sv.sv_id] = True
+    return svg
+
+  # For testing.
+  def json(self) -> dict:
+    return {
+        "svg_id": self.svg_id,
+        "svg_name": self.svg_name,
+        "sv_ids": list(self.sv_ids.keys()),
+        "parent_svg_ids": list(self.parent_svg_ids.keys()),
+        "child_svg_ids": list(self.child_svg_ids.keys())
+    }
+
+  def __str__(self) -> str:
+    return json.dumps(self.json(), indent=1)
 
-class StatVarPVs:
-  """An intermediate helper object that extracts SVs from input triples
-  and puts their PVs in a dict for quick lookup later.
-  It also maintains a separate dict for population types since SV hierarchies
-  are rooted at a group representing their population type.
-  """
 
-  def __init__(self, triples: list[Triple]) -> None:
-    self.sv_id_2_population_type: dict[str, str] = {}
-    self.sv_id_2_pvs: dict[str, dict[str, str]] = {}
+@dataclass
+class StatVarHierarchy:
+  svgs: dict[str, SVG]
+  svg_triples: list[Triple]
 
-    # Collect all triples into pop type and pv dictionaries.
-    # Also collect SV DCIDs for filtering subsequently.
 
-    # Using dict instead of set to maintain order.
-    # Maintaining order maintains results consistency and helps with tests.
-    sv_ids: dict[str, bool] = {}
-    dcid2poptype: dict[str, str] = {}
-    dcid2pvs: dict[str, dict[str, str]] = {}
+def create_all_svgs(svs: list[SVPropVals]) -> dict[str, SVG]:
+  svgs = create_leaf_svgs(svs)
+  for svg_id in list(svgs.keys()):
+    create_parent_svgs(svg_id, svgs)
+  return svgs
 
-    for triple in triples:
-      value = triple.object_id or triple.object_value
-      if not value:
-        logging.warning("Skipping, no value found for triple (%s).",
-                        str(triple))
+
+# Create SVGs that the SVs are directly attached to.
+def create_leaf_svgs(svs: list[SVPropVals]) -> dict[str, SVG]:
+  svgs: dict[str, SVG] = {}
+  for sv in svs:
+    svg: SVG = SVG.from_sv(sv, insert_sv_id=True)
+    svgs[svg.svg_id] = svg
+  return svgs
+
+
+def create_parent_svg(parent_sv: SVPropVals, svg: SVG, svgs: dict[str, SVG],
+                      svg_has_prop_without_val: bool):
+  parent_svg_id = parent_sv.gen_svg_id()
+  parent_svg = svgs.setdefault(parent_svg_id, SVG.from_sv(parent_sv))
```

Comment:
Does this always construct an SVG?  instead can we do a if key exists and construct when it doesn't?